src/fastpath/fastpath.c

```diff
@@ -502,7 +731,52 @@ void NORETURN fastpath_reply_recv(word_t cptr, word_t msgInfo)
     callerSlot->cteMDBNode = nullMDBNode;
 #endif
 
-    /* I know there's no fault, so straight to the transfer. */
+#ifdef CONFIG_EXCEPTION_FASTPATH
+    if (unlikely(fault_type != seL4_Fault_NullFault)) {
+        bool_t restart = 0;
+        switch (fault_type) {
+        /* Add cases for other types of faults here */
+        default:
+            restart = 1;
+        }
+
+        /* Note - this works as is for VM faults but will need to be changed when other faults are added. This is because
+         * vm faults always restart the faulting thread upon reply but this is not necessarily the case with other types
+         * of faults, which may instead become inactive. This can either be handled in the slowpath or redirected to the
+         * slowpath, but either way it must be changed so that we do not try and forcefully switch to a thread which is
+         * meant to be inactive. */
+        if (restart) {
+            word_t pc = getRestartPC(caller);
+            setNextPC(caller, pc);
+        }
+
+        /* Clear the tcbFault variable to indicate that it has been handled. */
+        caller->tcbFault = seL4_Fault_NullFault_new();
+
+        /* Dest thread is set Running, but not queued. */
+        thread_state_ptr_set_tsType_np(&caller->tcbState, ThreadState_Running);
+        switchToThread_fp(caller, cap_pd, stored_hw_asid);
+
+        /* The badge/msginfo do not need to be not sent - this is not necessary for exceptions */
+        restore_user_context();
+    } else {
+        /* There's no fault, so straight to the transfer. */
+
+        /* Replies don't have a badge. */
+        badge = 0;
+
+        fastpath_copy_mrs(length, NODE_STATE(ksCurThread), caller);
+
+        /* Dest thread is set Running, but not queued. */
+        thread_state_ptr_set_tsType_np(&caller->tcbState, ThreadState_Running);
+        switchToThread_fp(caller, cap_pd, stored_hw_asid);
+
+        msgInfo = wordFromMessageInfo(seL4_MessageInfo_set_capsUnwrapped(info, 0));
+
+        fastpath_restore(badge, msgInfo, NODE_STATE(ksCurThread));
+    }
+#else
```

Comment:
Is it possible to use the preprocessor defines to avoid the duplication of code here and conditionally place the original into an else branch?  Something like: 
```suggestion
    } else {
#else
```

Reply:
And then at replace the `#endif` below with:
```
#ifdef CONFIG_EXCEPTION_FASTPATH
    }
#endif
```